pattern.c

```diff
@@ -1621,12 +1633,16 @@ struct Pattern *mutt_pattern_comp(/* const */ char *s, int flags, struct Buffer
  * @param cache Cached Patterns
  * @retval true If ALL of the Patterns evaluates to true
  */
-static bool perform_and(struct Pattern *pat, enum PatternExecFlag flags,
+static bool perform_and(struct PatternHead *pat, enum PatternExecFlag flags,
                         struct Mailbox *m, struct Email *e, struct PatternCache *cache)
 {
-  for (; pat; pat = pat->next)
-    if (mutt_pattern_exec(pat, flags, m, e, cache) <= 0)
+  struct PatternHead p = *pat;
+  while (!SLIST_EMPTY(&p))
+  {
+    if (mutt_pattern_exec(&p, flags, m, e, cache) <= 0)
       return false;
+    SLIST_REMOVE_HEAD(&p, entries);
+  }
   return true;
```

Comment:
After force push the commit that is commented is gone, so here is the snippet:

```c
static int perform_or(struct PatternHead *pat, enum PatternExecFlag flags,
                      struct Mailbox *m, struct Email *e, struct PatternCache *cache)
{
  struct PatternHead p = *pat;
  while (!SLIST_EMPTY(&p))
  {
    if (mutt_pattern_exec(&p, flags, m, e, cache) > 0)
      return true;
    SLIST_REMOVE_HEAD(&p, entries);
  }
  return false;
}
```

The original function would traverse the list and pass the sublist to `mutt_pattern_exec`:
```c
for (; pat; pat = pat->next)
  if (mutt_pattern_exec(pat, flags, m, e, cache) > 0)
  ...
```

The code I submit emulates the same functionality. This cannot be done with `SLIST_FOREACH()` since it iterates over the list values and cannot be used to construct sub-lists which are then passed to `mutt_pattern_exec`.

Reply:
If I read the function correctly, `mutt_pattern_exec` uses only the head of the linked list, which it extracts with `struct Pattern *first = SLIST_FIRST(pat);`. How about you iterate with a `SLIST_FOREACH()` as @flatcap suggests and you just pass down the node to work on?

Reply:
Yeah makes sense to make `mutt_pattern_exec` accept a `Pattern` instead of a `PatternHead` since it only operates on the first (head) value. I've made the changes so that now perform_and/or will use `SLIST_FOREACH`.